src/kernel/src/random/mod.rs

```diff
@@ -0,0 +1,197 @@
+    let res = acc.borrow_mut().try_fill_random_data(out);
+    if let Ok(()) = res {
+        return true;
+    }
+    logln!("need to seed accumulator");
+    // try_fill_random_data only fails if unseeded
+    // so the rest is trying to seed it/wait for it to be seeded
+    let mut entropy_sources = ENTROPY_SOURCES
+        .call_once(|| Mutex::new(EntropySources::new()))
+        .lock();
+    if entropy_sources.has_sources() {
+        logln!("has sources");
+        entropy_sources.contribute_entropy(acc.borrow_mut());
+        acc.try_fill_random_data(out)
+            .expect("Should be seeded now & therefore shouldn't return an error");
+        drop((entropy_sources, acc));
+        return getrandom(out, nonblocking);
+    }
+    drop((entropy_sources, acc));
+    if nonblocking {
+        // doesn't block, returns false instead
+        false
+    } else {
+        // otherwise schedule and recurse in again after this thread gets picked up again
+        // this way it allows other work to get done, work that might result in entropy events
+
+        // block for 2 seconds and hope for other entropy-generating work to get done in the
+        // meantime
+        logln!("recursing");
+        sys_thread_sync(&mut [], Some(&mut Duration::from_secs(2))).expect(
+            "shouldn't panic because sys_thread_sync doesn't panic if no ops are passed in",
+        );
+        getrandom(out, nonblocking)
+    }
+}
+
+/// Be sure to contribute at least one byte and at most 32 bytes.
+pub fn contribute_entropy(
+    contributor: &mut Contributor,
+    event: &[u8],
+) -> Result<(), self::fortuna::Error> {
+    let mut acc = ACCUMULATOR
+        .call_once(|| Mutex::new(Accumulator::new()))
+        .lock();
+    acc.add_random_event(contributor, event)
+}
+/// Returns whether registration was successful
+pub fn register_entropy_source<T: EntropySource + 'static + Send + Sync>() -> bool {
+    // ignore whether or not the registration was successful
+    let mut entropy_sources = ENTROPY_SOURCES
+        .call_once(|| Mutex::new(EntropySources::new()))
+        .lock();
+    let res: Result<(), ()> = entropy_sources.try_register_source::<T>();
+    res.is_ok()
+}
+
+pub fn start_entropy_contribution_thread() {
+    // let thread = start_new_kernel(
+    //     Priority::default_background(),
+    //     contribute_entropy_regularly,
+    //     0,
+    // );
+    let _registered_cpu = maybe_add_cpu_entropy_source();
+    let _registered_jitter = maybe_add_jitter_entropy_source();
+    // FIXME: currently this thread never is actually run again due to
+    // default_background priority coupled with sys_thread_sync never actually
+    // causing the thread to resume.
+    run_closure_in_new_thread(Priority::default_background(), || {
+        contribute_entropy_regularly()
+    });
+}
+
+extern "C" fn contribute_entropy_regularly() {
+    logln!("Starting entropy contribution");
+    loop {
+        logln!("Contributing entropy");
```

Comment:
Is this going to print every second? In general, maybe reduce some of the loglns in this file?